agent/eventhandler/attachment_handler.go

```diff
@@ -0,0 +1,150 @@
+		return fmt.Errorf("eventhandler: attachment handler received unrecognized event type: %d", change.GetEventType())
+	}
+	event, ok := change.(api.AttachmentStateChange)
+	if !ok {
+		return errors.New("eventhandler: unable to get attachment event from state change event")
+	}
+
+	if event.Attachment == nil {
+		return fmt.Errorf("eventhandler: received malformed attachment state change event: %v", event)
+	}
+
+	attachmentARN := event.Attachment.AttachmentARN
+	handler.lock.Lock()
+	if _, ok := handler.attachmentARNToHandler[attachmentARN]; !ok {
+		handler.attachmentARNToHandler[attachmentARN] = &attachmentHandler{
+			attachmentARN: attachmentARN,
+			stateSaver:    handler.stateSaver,
+			client:        handler.client,
+			ctx:           handler.ctx,
+		}
+	}
+	handler.lock.Unlock()
+
+	attachmentHandler := handler.attachmentARNToHandler[attachmentARN]
+	go attachmentHandler.submitAttachmentEvent(&event)
+
+	return nil
+}
+
+// submitAttachmentEvent submits an attachment event to backend
+func (handler *attachmentHandler) submitAttachmentEvent(attachmentChange *api.AttachmentStateChange) {
+	seelog.Debugf("AttachmentHandler: acquiring attachment lock before sending attachment state change for attachment %s", handler.attachmentARN)
+	handler.lock.Lock()
+	seelog.Debugf("AttachmentHandler: acquired attachment lock for attachment %s", handler.attachmentARN)
+	defer handler.lock.Unlock()
+
+	backoff := GetSubmitAttachmentStateBackoffFunc()
+
+	retry.RetryWithBackoffCtx(handler.ctx, backoff, func() error {
+		if !attachmentChangeShouldBeSent(attachmentChange) {
+			seelog.Debugf("AttachmentHandler: not sending attachment state change [%s] as it should not be sent", attachmentChange.String())
+			// if the attachment state change should not be sent, we don't need to retry anymore so return nil here
+			return nil
+		}
+
+		seelog.Infof("AttachmentHandler: sending attachment state change: %s", attachmentChange.String())
+		if err := handler.client.SubmitAttachmentStateChange(*attachmentChange); err != nil {
+			seelog.Errorf("AttachmentHandler: error submitting attachment state change [%s]: %v", attachmentChange.String(), err)
+			return err
+		}
+
+		attachmentChange.Attachment.SetSentStatus()
+		attachmentChange.Attachment.StopAckTimer()
+		handler.stateSaver.Save()
+		seelog.Debugf("AttachmentHandler: submitted attachment state change: %s", attachmentChange.String())
+		return nil
+	})
+}
+
+// attachmentChangeShouldBeSent checks whether an attachment state change should be sent to backend
+func attachmentChangeShouldBeSent(attachmentChange *api.AttachmentStateChange) bool {
+	return !attachmentChange.Attachment.HasExpired() && !attachmentChange.Attachment.IsSent()
+}
+
+// GetSubmitAttachmentStateBackoffFunc is a func that returns a backoff object;
+// this is made as a function object so that we can use smaller backoff in unit test
+var GetSubmitAttachmentStateBackoffFunc = GetSubmitAttachmentStateBackoff
```

Comment:
a. does this need to be a global? 
b. can you please move all var's and const's to the top of the file?
c. it seems like you're using this to override in your tests. That's an anti-pattern as well. Having global variables that get overridden in tests leads to flaky tests. Can you make that a part of the `attachmentHandler` struct or something and override it in the test package when you construct the object?

Reply:
made the backoff as a field of attachmentHandler and overridden it in test